gensim/models/poincare.py

```diff
@@ -731,27 +739,457 @@ def compute_loss(self):
+            Maximum number of descendants to return.
+
+        Returns
+        -------
+        List (str)
+            Descendant nodes from the node `term_1`.
+
+        """
+        depth = 0
+        descendants = []
+        current_node = term_1
+        while depth < max_depth:
+            descendants.append(self.closest_child(current_node))
+            current_node = descendants[-1]
+            depth += 1
+        return descendants
+
+    def ancestors(self, term_1):
+        """
+        Returns the list of recursively closest parents from the given node.
+
+        Parameters
+        ----------
+        term_1 : str
+            Input term.
+
+        Returns
+        -------
+        List (str)
+
+        """
+        ancestors = []
+        current_node = term_1
+        ancestor = self.closest_parent(current_node)
+        while ancestor is not None:
+            ancestors.append(ancestor)
+            ancestor = self.closest_parent(ancestors[-1])
+        return ancestors
+
+    def cluster(self, breaks):
+        all_norms = np.linalg.norm(self.syn0, axis=1)
+        clusters = []
+        breaks[-1] = 1.0
+        for i in range(len(breaks) - 1):
+            lower_bound = breaks[i]
+            upper_bound = breaks[i+1]
+            cluster_indices = np.where((all_norms >= lower_bound) & (all_norms < upper_bound))[0]
+            clusters.append([self.index2word[index] for index in cluster_indices])
+        return clusters
+
+    def distance(self, term_1, term_2):
+        """
+        Return Poincare distance between two terms.
+
+        Parameters
+        ----------
+        term_1 : str
+            Input term.
+        term_2 : str
+            Input term.
+
+        Returns
+        -------
+        float
+            Poincare distance between the vectors for `term_1` and `term_2`.
+
+        Examples
+        --------
+
+        >>> model.distance('mammal.n.01', 'carnivore.n.01')
+        2.13
+
+        Notes
+        -----
+        Raises KeyError if either of `term_1` and `term_2` is absent from vocab.
+
+        """
+        vector_1 = self.word_vec(term_1)
+        vector_2 = self.word_vec(term_2)
+        return self.poincare_distance(vector_1, vector_2)
+
+    def similarity(self, term_1, term_2):
+        """
+        Return similarity based on Poincare distance between two terms.
+
+        Parameters
+        ----------
+        term_1 : str
+            Input term.
+        term_2 : str
+            Input term.
+
+        Returns
+        -------
+        float
+            Similarity between the vectors for `term_1` and `term_2` (between 0 and 1).
+
+        Examples
+        --------
+
+        >>> model.similarity('mammal.n.01', 'carnivore.n.01')
+        0.73
+
+        Notes
+        -----
+        Raises KeyError if either of `term_1` and `term_2` is absent from vocab.
+        Similarity lies between 0 and 1.
+
+        """
+        return 1 / (1 + self.distance(term_1, term_2))
```

Comment:
IIRC, we wanted to normalize the distance (`norm_distance = distance / max_distance`) and then return `1 - norm_distance` as similarity. I'm re-thinking it now. One feature would be that the similarity would be 0 for the max distance. I'm not sure if that's desired. Maybe the current is OK.

Reply:
Yes, keeping this in the current form, due to the significant overhead of computing max distance.